### ISSUE TITLE
Cross-platform, solutions and answers

### ISSUE THREAD

**Issue by EvgenyPrikhodko** [dotnet/runtime#98123]
### Background and motivation

The programmer may not know about the operation of strings

### API Proposal

*

### API Usage

```csharp
   str.AppendLine("test1");
        str.AppendLine("test2");
        str.AppendLine("test3");

        var qryColumnsString = string.Join(", ", str.ToString().Remove(str.Length - 2).Split("\n"));
```

### Alternative Designs

cross-platform

**Comment by CyrusNajmabadi** [comment]
@EvgenyPrikhodko As many would not know the connection between this issue and hte last one raised, could you please flesh this out more to indicate:

1. what the problem is that you would like a solution for.
2. what the proposed solution is here that you're looking for.

**Comment by tannergooding** [comment]
Yes, ideally this would cover the root issue that `StringBuilder.Append` exclusively uses `Environment.NewLine` and that users may be trivially unaware of this, so if they do `sb.Split('\n')` or `sb.Split("\r\n")` which can differ on some platforms (miss splitting the `\r` on Windows and miss all lines on Unix).

Then explain that the proposal is to add an analyzer which likes for such patterns and recommends using `Environment.NewLine` instead.

It should acknowledge that there is the potential for `false positives` (user did want to split on just the specified newline, even if different from the platform default) and for `false negatives` (user was unaware of the string shape, but was using the right API, so changing would break).

**Comment by EvgenyPrikhodko** [comment]
the programmer is waiting for a lonely answer when we talk about different operating systems

**Comment by EvgenyPrikhodko** [comment]
@tannergooding critical error

**Comment by CyrusNajmabadi** [comment]
@EvgenyPrikhodko Your posts are quite cryptic.  Could you please provide more clarifying details when you make them?  Thanks! :)

**Comment by EvgenyPrikhodko** [comment]
> @EvgenyPrikhodko Your posts are quite cryptic. Could you please provide more clarifying details when you make them? Thanks! :)

![image](https://github.com/dotnet/runtime/assets/25795145/f079a333-50d5-4c84-ae74-8a61cfd65d1e)

**Comment by EvgenyPrikhodko** [comment]
> @EvgenyPrikhodko Your posts are quite cryptic. Could you please provide more clarifying details when you make them? Thanks! :)

![image](https://github.com/dotnet/runtime/assets/25795145/a32da25d-68ff-4b2c-ae0b-97443f2f7044)

**Comment by CyrusNajmabadi** [comment]
@EvgenyPrikhodko i'm referring to https://github.com/dotnet/runtime/issues/98123#issuecomment-1932832824 and https://github.com/dotnet/runtime/issues/98123#issuecomment-1932835418.

You have also not clearly and reasonably answered: https://github.com/dotnet/runtime/issues/98123#issuecomment-1932812190

**Comment by Clockwork-Muse** [comment]
... part of the problem here is that generally you _shouldn't_ be manually splitting lines via `string.Split()`.  For most simple cases, a `StringReader` is the better choice.  That's assuming you can't read lines from whatever originating stream (such as directly reading lines from a file). 
This is especially important if you're parsing a file, because you don't actually have a guarantee about the line endings in the file (outside of some custom file format, I suppose, but you shouldn't be using those if you could help it).  Especially for interoperability concerns, if you're parsing text files you need to assume that you will encounter either style.  So while we could(/should?) write an analyzer for flagging on platform specific `string.Split("\r\n")`, doing `string.Split(Environment.NewLine)` _is liable to fail_.

The same holds true when writing lines as well, to an extent.  You shouldn't be using `string.Join()` to concatenate lines, instead using `StringBuilder.AppendLine()` or using `StreamWriter.WriteLine()`.

**Comment by EvgenyPrikhodko** [comment]
> @EvgenyPrikhodko As many would not know the connection between this issue and hte last one raised, could you please flesh this out more to indicate:
> 
> 1. what the problem is that you would like a solution for.
> 2. what the proposed solution is here that you're looking for.

1. Programmer may come to an existing project and not notice when there is a lot of code that somewhere in the solution there is a StringBuilder that is platform-dependent. 

2. Net is positioned as a cross-platform solution, or Net gives a warning, or solves similar problems on its own. But, let's have different results in different operating systems — this is bad.

**Comment by CyrusNajmabadi** [comment]
> > @EvgenyPrikhodko As many would not know the connection between this issue and hte last one raised, could you please flesh this out more to indicate:
> > 
> > 1. what the problem is that you would like a solution for.
> > 2. what the proposed solution is here that you're looking for.
> 
> 1. Programmer may come to an existing project and not notice when there is a lot of code that somewhere in the solution there is a StringBuilder that is platform-dependent. 
> 
> 2. Net is positioned as a cross-platform solution, or Net gives a warning, or solves problems on its own.  But, get different results in different operating systems — this is bad.
> 

Thanks. Can you please update the op?
